removed for loop, changed arguments of convert()

markvrma opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/alan-turing-institute/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Fixes #1856 

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
- Made changes to `_tags`, removed `for loop` to remove inefficient vectorization.
#### Does your contribution introduce a new dependency? If yes, which one?
No
<!--
If your contribution does add a new hard dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum.
-->

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->
- [ ] running `pytest` on `test_dwt.py` gives tabularization errors. I'm not sure how to handle this. Out of 11 tests, 6 give the same error
```
================================================================== short test summary info ==================================================================
FAILED test_dwt.py::test_output_of_transformer - ValueError: Tabularization failed, it's possible that not all series were of equal length
FAILED test_dwt.py::test_no_levels_does_no_change - ValueError: Tabularization failed, it's possible that not all series were of equal length
FAILED test_dwt.py::test_output_dimensions[2-12] - ValueError: Tabularization failed, it's possible that not all series were of equal length
FAILED test_dwt.py::test_output_dimensions[3-11] - ValueError: Tabularization failed, it's possible that not all series were of equal length
FAILED test_dwt.py::test_output_dimensions[4-12] - ValueError: Tabularization failed, it's possible that not all series were of equal length
FAILED test_dwt.py::test_dwt_performs_correcly_along_each_dim - ValueError: Tabularization failed, it's possible that not all series were of equal length
```
- Will it be right to use [padder.py](https://github.com/alan-turing-institute/sktime/blob/main/sktime/transformations/panel/padder.py) to make all series of equal length? 
     
- [ ] I haven't moved the file to `series` folder yet, once the error is fixed, it can be moved along with the `test_dwt`.
#### Any other comments?
- I'm not sure if my approach is correct, please let me know if there's a better way to approach this. 
- Fixed a typo in `_convert.py`
- I have commented out code for my reference rather than deleting them, once the error is fixed I'll delete them.
- I haven't moved the file to `series` folder yet, once the error is fixed, it can be moved along with the `test_dwt`.
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

fkiraly commented:
@markvrma, thanks for your contribution!

I´m not quite sure what you are trying to do, would you mind explaining?

markvrma commented:
The intent was to try and apply the 1st method as mentioned in the #1856. 

- modified the `_tags`, `"X_inner_mtype": "pd-multiindex"`
- removed the `for loop` to eliminate inefficient vectorisation (though I'm not sure if I should've done that)
- I could not find a way to convert `from_type = pd-multiindex` in the `convert` function to a `1D array` / `numpyflat` hence I stuck with `nested_univ`

fkiraly reviewed (changes requested):
I see, thanks for clarifying!

My recommendation would be:
* use the `numpy3D` internally, i.e., `X_inner_mtype="numpy3D"`, and use `numpy` vectorization and reshape.
* ensure you also return a `numpy3D` array.

That´s because the `_extract_wavelet_coefficients` utility seems to be using `numpy` arrays internally.
Would that work?

markvrma commented:
> * ensure you also return a `numpy3D` array.

 Will it be a problem since tests in `test_dwt.py` rely on `_transform` returning a nested pandas DataFrame?

fkiraly commented:
>Will it be a problem since tests in test_dwt.py rely on _transform returning a nested pandas DataFrame?

No, because it tests `transform` and not `_transform`, and the former should convert back to a nested pandas df if one is passed.

markvrma commented:
- converts from `numpy3D` to `numpyflat` back to `numpy3D`
- `check_if_dataframes_are_equal` throws AssertionError (fails 3 tests)
   - Upon closer inspection( printing `df1` and `df2`  in the  `check_if_dataframes_are_equal` function ) it looks like the column names of `df1` and `df2` are different and for the 3rd test there is only 1 column in `df1` while `df2` has 2 columns

fkiraly commented on sktime/transformations/panel/dwt.py:
```diff
@@ -23,12 +24,12 @@ class DWTTransformer(BaseTransformer):
 
     _tags = {
-        "scitype:transform-input": "Series",
+        "scitype:transform-input": "Panel",
```
this should be `Series`

fkiraly commented:
>  3rd test there is only 1 column in df1 while df2 has 2 columns

Yes, you removed the loop over columns. I still think you may need to loop over the variables (2nd index in the numpy 3D container)
